### ISSUE TITLE
Consider removing explicit freeze period in TicketBroker

### ISSUE THREAD

**Issue by yondonfu** [livepeer/protocol#306]
At the moment, the TicketBroker contract explicitly defines a `freezePeriod` parameter. The value of this parameter is the # of rounds that a sender's reserve remains frozen after the sender overspends from its deposit and a recipient claims from the reserve. 

There are a few downsides with the way the explicit multi-round freeze period is currently implemented:

- More state tracking costs. The contract stores the freeze round, the # of recipients in the freeze round and a "reserve nonce" that is used to incremented whenever a reserve is funded after it was previously frozen (the nonce is used to point to a new mapping that tracks amounts claimed by recipients when a reserve is frozen). The main transactions that these costs are passed on to are reserve claiming ticket redemptions and funding a reserve after it was previously frozen
- More client complexity. Recipients might need to handle frozen vs. non-frozen reserves differently. For example, with an explicit multi-round freeze period, if a recipient encounters a frozen reserve it would need to check if it was in the active set in the freeze round as opposed to just checking if it is in the active set in the current round. Furthermore, recipients need to make a decision about whether to work with a sender with a frozen reserve. If the decision is to work with a sender with a frozen reserve then the client needs to also make sure that the reserve is not thawed (check if current round >= the reserve's thaw round)
- Does not allow a sender top off its reserve immediately after the reserve is frozen - the sender needs to wait through the multi-round freeze period first. This UX is a bit rough and treats overspending fairly harshly when in reality it could happen by accident.

An alternative to consider is to remove the explicit freeze period and instead if overspending happens the reserve only stays committed to current round's recipient set. Essentially, the only thing different about the overspending scenario is that funds come from the reserve instead of the deposit and there is a cap on the claimable amount for each recipient. The main downside of this alternative is that if a sender overspends at the very end of a round and a recipient is evicted from the active set as the new round begins, then the recipient will not be able to redeem any received winning tickets to claim from the sender's reserve. The explicit multi-round freeze period helps with this because if someone freezes the sender's reserve at the end of the round, then all recipients are still guaranteed the ability to redeem received winning tickets to claim from the sender's reserve in the following round. However, the benefit of the freeze period here depends on someone actually freezing the sender's reserve before the round ends - if no one freezes the reserve in time then any evicted recipients in the next round would still be out of luck. It is also worth noting that being evicted from the active set in a following round is a general problem even without overspending - if a recipient is evicted in the next round then it will be unable to redeem any received winning tickets.

I think this alternative has a few benefits:

- Less state tracking costs. The contract does not need to store any state pertaining to the freeze round or # of recipients in the freeze round - if the reserve is only frozen for the remainder of the round, then the contract can just rely on contract queries that will be accurate for the round. The "reserve nonce" does not need to be stored either - instead we can use the current round number as a pointer to new mappings that track amounts claimed from a reserve in each round
- Less client complexity. There would be no special casing required for frozen vs. non-frozen reserves. Recipients would always just check if they are active in the current round and the size of the active set in the current round. Recipients would not need to decide whether to work with senders with frozen vs. non-frozen reserves - recipients can work with these senders as long as the allocation from the reserve is sufficient
- Allow a sender to top off its reserve immediately after it is frozen. This treats overspending less harshly - if it happens a sender can quickly top off its funds and continue paying recipients

At a high level, the implementation change for this alternative would:

- Use round numbers to point at mappings for claimed amounts by recipients
- If overspending happens, record the claimable amount at the time of overspending for the round so that value can be used to calculate reserve allocations
- Use a single state variable to track remaining funds in a reserve
- If funds are added to a reserve during a round when overspending happened, update the funds in the reserve, but also the claimable amount for the reserve (to allow recipients in the round to access those new funds if they claim from the reserve)

Most of these changes are outlined in [this spike](https://github.com/livepeer/protocol/compare/pm...yf/freeze-reserve-remainder-of-round-spike). Note: the spike branch still uses the frozen vs. non-frozen terminology for now.

**Comment by yondonfu** [comment]
Made a few updates to [this spike](https://github.com/livepeer/protocol/compare/pm...yf/freeze-reserve-remainder-of-round-spike):

- Removed reserve states. We don't need to differentiate between a frozen vs. non-frozen reserve when someone claims from a reserve. We can just keep track of reserve allocations that each member of the active set is entitled to based on the funds remaining in the reserve, the funds claimed from the reserve in the current round and the funds claimed by an address during the current round.

During round N, the formula for calculating the current reserve allocation for an address A that is in the active set is:

```
((reserve.funds + reserve.claimedForRound[N]) / numRecipients) - reserve.claimedByAddress[A]
```

- `reserve.funds` decreases whenever funds are claimed from the reserve. However, even as funds are claimed, since we keep track of the amount claimed the round N, we can always sum `reserve.funds` and `reserve.claimedForRound[N]` to derive the value that reserve allocations should be calculated based off

- Whenever X is claimed from a reserve in round N by address A, increase `reserve.claimedForRound[N]` by X and increase `reserve.claimedByAddress[A]` by X.

**Comment by yondonfu** [comment]
Closed by #319
